### ISSUE TITLE
Hillshading for Mapsforge

### ISSUE THREAD

**Issue by ztNFny** [cgeo/cgeo#12210]
### Tell us your idea!

c:geo should offer hillshading on (offline) maps which is the coloring of the map to show elevation changes.

Locus and OruxMaps both do that on top of mapsforge offline vector maps, so it is available and working, but I'm having a hard time finding documentation on it.

![image](https://user-images.githubusercontent.com/1258173/144286497-6e3de5dc-edcd-4960-8053-5a65af77e57a.png)

### Additional Information

https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map/src/main/java/org/mapsforge/map/rendertheme/renderinstruction/Hillshading.java
https://docs.locusmap.eu/doku.php?id=manual:faq:how_to_add_map_shading#where_to_get_them
hgt-source: https://bit.ly/dtm-germany-1s

### Subtasks

- [x] Make it work
- [x] Integrate with SAF
- [ ] Add download manager (either like Routing or Maps)

**Comment by moving-bits** [comment]
There are some samples for Android apps under https://github.com/mapsforge/mapsforge/tree/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android, and you can contact the devs via their mailing list mapsforge-dev@googlegroups.com

**Comment by ztNFny** [comment]
I've started implementing this in https://github.com/ztNFny/cgeo/tree/issue/12210_hillshading

The major obstacle so far is SAF. Mapsforge requires a `java.io.File` object for the folder containing the dem files but as far as I understand with SAF that's not possible.
As a workaround to get the actual functionality working I've added `android.permission.MANAGE_EXTERNAL_STORAGE`, but obviously that can't stay.

@moving-bits looking at PRs it seems you've implemented most SAF stuff in c:geo. Is it true that getting a File isn't possible or could you give me a pointer as to how?
If it isn't possible I guess we'd have to re-implement all classes in `org.mapsforge.map.layer.hills` (in mapsforge upstream of course)

**Comment by moving-bits** [comment]
> @moving-bits looking at PRs it seems you've implemented most SAF stuff in c:geo.

99% of SAF was implemented by @eddiemuc actually.

He also added some PR to mapsforge to help them become SAF-compatible for theming. Maybe this would be the right way as well for hill shading?

> Is it true that getting a File isn't possible or could you give me a pointer as to how?

IIRC getting `File` objects is possible only for non-public folders, meaning for the app-specific Android-internal folder. This would be no option if we want to share such data between apps.

BTW: Isn't hill shading data already included in the map files available from openandromaps.org?

**Comment by eddiemuc** [comment]
Hi @ztNFny 

There is no way to get a java.io.File object via SAF. This is the key point which made the whole SAF migration so much work :-)

Mapsforge supports already opening map offline files via InputStreams (rather than File objects) which we use today to overcome the SAF obstacles. For Map Themes, however, we had a similar challenge to solve. This was not possible without adding functionality to mapsforge itself back in the days when I did this.

There are multiple ways we could go:
* First question: is it necessary / meaningful that the folder to use is public? If not then we can place it in the private app space - this space can still be accessed using File objects as ever.
     * (I currently don't know anything about the files in question. Are they large or small? Are they generated by mapsforge itself (and the folder is just a caching place) or are they downloaded from some other source?)
* If they need to be in a public space: could they be placed in one big ZipFile which we load into memory and source mapsforge from there? (This is actually one way I enabled fpr map themes. Changes inside mapsforge were necessary for this though(
* If this is also not possible: mapsforge might be reworkable to use InputStreams rather than File objects in this area. However, unfortunately live access through SAF is horribly slow! (This is why as a second option for map themes I implemented synchronizing map theme files from public storage into app private storage and use files from there)

Can you provide a little more input about the nature of the files in question? Esp typical size and where they are originating from?

**Comment by ztNFny** [comment]
> BTW: Isn't hill shading data already included in the map files available from openandromaps.org?

No, they only include the info for contour lines (which may or may not be shown depending on your mapsforge theme), not the shading itself. But the source for generating those contour lines is the same as used for hillshading: http://data.opendataportal.at/dataset/dtm-germany

**Comment by ztNFny** [comment]
> is it necessary / meaningful that the folder to use is public?

Like maps they can be shared between apps, so I'd say yes

> I currently don't know anything about the files in question. Are they large or small? Are they generated by mapsforge itself (and the folder is just a caching place) or are they downloaded from some other source?

Germany is 1.7 GB, downloaded from http://data.opendataportal.at/dataset/dtm-germany

Direct download link is https://bit.ly/dtm-germany-1s - I think that answers the other questions too

**Comment by eddiemuc** [comment]
Totally unrelated question: I noticed in the header of this issue a notice "1 of 3 tasks":

![image](https://user-images.githubusercontent.com/6909759/144718412-9a73a5b0-26a2-4460-9568-c6dbdd2ac45e.png)

What is this? It looks somehow useful, but what does it relate to?

**Comment by ztNFny** [comment]
> What is this? It looks somehow useful, but what does it relate to?

In the original issue you can define subtasks: "- [ ] task 1" resp "- [x] completed task"

**Comment by eddiemuc** [comment]
Ok, I did a very quick scan of the mapsforge code involved. So far I found the following:

The code is part of the generic mapsforce map code (not in the android specific part), and it was written around 2017 (so well before the whole SAF stuff came up). 

I identified two main parts using the "file" information:

* In class HgtCache there is a crawl method which scans the given folder recursively for tile files matching a certain name pattern. It builds up sort of an "index map" from tile keys to hill data files. It also opens and scans Zip files in this process => this code part uses File functions like "isDirectory/isFile/listFiles" which is not available in SAF. This is tricky to replace
* In class AbsShadingAlgorithmDefaults the file contents are actually read and processed => File objects are only used there to open an InputStream, this could kind of easy be changed to SAF semantics

Fazit so far:
* I don't see any way to enable this together with SAF without investing substancial effort in the mapsforge code
* Since the Hill stuff is in the Non-Android-specific part, implementing this must abstract over SAF too - because the code needs to be still usable in Non-Android contexts. So a bit of architectural/design work would be necessary
* The mapsforge owners have to agree to such a huge change. If this way should be tried, we should contact them beforehand and start discussion (not just coding ahead...)
* We will at some point face the "SAF-very-very-slow" problem here again (as we did with map themes), I have no good solution for this so far. Keeping data in-memory will certainly not work here...

For reference: HgtCache.crawl function in mapsforge:
https://github.com/mapsforge/mapsforge/blob/e558dea5e9a22d5c9b500d01c2a6b9b92b100643/mapsforge-map/src/main/java/org/mapsforge/map/layer/hills/HgtCache.java#L159

For reference: AbsShadingAlgorithmDefaults, point of first usage of the File reference:
https://github.com/mapsforge/mapsforge/blob/e558dea5e9a22d5c9b500d01c2a6b9b92b100643/mapsforge-map/src/main/java/org/mapsforge/map/layer/hills/AbsShadingAlgorithmDefaults.java#L43

TL;DR: this is goinfg to be a lot of work.

**Comment by ztNFny** [comment]
I'll open an issue @ mapsforge, let's see...

**Comment by ztNFny** [comment]
Their response doesn't make it seem that they care too much ... https://github.com/mapsforge/mapsforge/discussions/1269#discussioncomment-1751800

**Comment by eddiemuc** [comment]
> Their response doesn't make it seem that they care too much ... https://github.com/mapsforge/mapsforge/discussions/1269#discussioncomment-1751800

I find the answer promising, since it looks like they would allow it to be included in their stack - if it is implemented Cross-Platform.

As far as I know mapsforge is not in the same comfortable position as c:geo in that they have only very few core developers (I myself only had contact with a single one). Major changes thus have to be contributed by the community, they might have just enough capacity to govern these changes and make sure the overall stack continues to work (on all supported platforms and systems, which is far far more than just Android).

**Comment by fm-sys** [comment]
If this gets implemented, can we also (optionally) show a altitude cursor like in locus?

![Screenshot_20211206_191320_menion android locus pro computerBild](https://user-images.githubusercontent.com/64581222/144902271-0bd0e0e0-050c-4584-962c-21144fbe6453.png)

**Comment by ztNFny** [comment]
> If this gets implemented, can we also (optionally) show a altitude cursor like in locus?

Not with the hillshading implementation in mapsforge - is really just shading. Basically it's adding a layer with shades of grey on top of the original map. 

https://github.com/ztNFny/cgeo/tree/issue/12210_hillshading contains a more or less complete hillshading implementation already, all that's missing is SAF which needs to be implemented upstream - if you want to take a look at how it works and what's avialable.

![image](https://user-images.githubusercontent.com/1258173/144909222-82ecf8fe-3388-435d-9a9d-5cae6c6bc3f9.png)

**Comment by Lineflyer** [comment]
Out of curiosity:
What is the source of such hillshading  information or is it already contained in the map file itself and generated using the change of elevation info from the map?

**Comment by ztNFny** [comment]
https://wiki.openstreetmap.org/wiki/SRTM

**Comment by ztNFny** [comment]
reading #12329 is it correct that for files in /sdcard/Android/data/ java.io.File can still be used? So if we stored the dtm files there then it'd work without having to implement SAF in upstream Mapsforge?

**Comment by eddiemuc** [comment]
> reading #12329 is it correct that for files in /sdcard/Android/data/ java.io.File can still be used? So if we stored the dtm files there then it'd work without having to implement SAF in upstream Mapsforge?

To the best of my knowledge this should not be possible according to Android standard. This does not mean that some editions may still allow this.

**Comment by fm-sys** [comment]
AFAIK SAF is only enforced for external non-app-speciffic locations / directories. So it should work. But not completely sure though...

**Comment by ztNFny** [comment]
I've tested it on Android 11 and 12: Placing the files in /storage/emulated/0/Android/data/cgeo.geocaching/files/ I can access them using java.io.File
Unless anyone is opposed to doing that I'd like to go ahead with that location for a v1.0 implementation.

**Comment by Lineflyer** [comment]
We should first of all align the nature of this data?
IMHO the mentioned folder is only suitable for files not needed to be access or modified by users (which also means we need to incorporate a download method within c:geo?).

Please also see here more information about the existing handled storage locations in c:geo: https://github.com/cgeo/cgeo/wiki/Disk-usage-structure

So going by that list you suggest place 2 (respectively 5 if the user moved the files to SD using c:geo settings). Correct?

**Comment by ztNFny** [comment]
I've used LocalStorage.getExternalPrivateCgeoDirectory() for my tests now. That folder is usually accessible with file managers for the user to put the files, but my aim would be to have a download function integrated - but this has some further stuff to clarify, like permissions to access those sources directly - so the download function will add a lot of complexity, not just in the technical sense.

**Comment by eddiemuc** [comment]
Hi @ztNFny, accessing the private file stores of an app is still possible using files.  I can definitely confirm that. SAF has to be used for files which should be shared across apps though.

**Comment by SammysHP** [comment]
> That folder is usually accessible with file managers for the user to put the files

Only on some versions of Android 12. In the future access to /sdcard/Android/data will be blocked. The only reason you can still operate with file handles on this directory is because nobody else has access to it.

**Comment by ztNFny** [comment]
> Only on some versions of Android 12. In the future access to /sdcard/Android/data will be blocked. The only reason you can still operate with file handles on this directory is because nobody else has access to it.

It'll even be blocked for android.permission.MANAGE_EXTERNAL_STORAGE? File managers will retain that permission.

**Comment by SammysHP** [comment]
> It'll even be blocked for android.permission.MANAGE_EXTERNAL_STORAGE?

AFAIK yes.

**Comment by fm-sys** [comment]
> 

- Internal Private Directory `/data/data/`
- Internal Public Directory `/emulated/0/Android/data/` or `/sdcard/Android/data/` 

Why should they have a _public_ directory if the user cannot access it?

**Comment by eddiemuc** [comment]
> > 
> 
> - Internal Private Directory `/data/data/`
> - Internal Public Directory `/emulated/0/Android/data/` or `/sdcard/Android/data/` 
> 
> Why should they have a _public_ directory if the user cannot access it? 

There is no such thing as a "Internal public directory". Android differentiates between an internal and an external directory (one on internal storage, one on SD card), but both are considered app-private.

That said, it is true that some apps under some conditions may be able to access other apps external private dirs (on SD card). But this is definitely not a mechanism Android intends for file sharing. Specifically:
* Two "normal" apps (without special permission) are not able to access this content at the same time (e.g. locus and c:geo sharing map files via this mechanism is impossible)
* Files in this folder are considered as belonging to the app. This means e.g. that they are deleted if the app is uninstalled

Reference doxu:
* https://developer.android.com/training/data-storage
* https://developer.android.com/training/data-storage/app-specific

**Comment by ztNFny** [comment]
> But this is definitely not a mechanism Android intends for file sharing.

It's clear that sharing between 2 apps (e.g. cgeo and Locus) requires SAF - and SAF would clearly be the target for this in the long run.

I'm proposing to use /sdcard/Android/data for the initial Hillshading implementation as that
- wouldn't require any changes in upstream Mapsforge 
- the user can place the hgt files there manually while there isn't a downloader (which requires a some site where we are allowed to download from)

**Comment by eddiemuc** [comment]
> It's clear that sharing between 2 apps (e.g. cgeo and Locus) requires SAF - and SAF would clearly be the target for this in the long run.
> 
> I'm proposing to use /sdcard/Android/data for the initial Hillshading implementation

If an implementation is started, why not doing it the right way the first time and enable necessary support in mapsforge? If this is initially done based on a hack, then chances are it has to be redone in large parts for clean implementation later - which is a big waste of effort. (and, I might add, not a job we will easily find a volunteer for later on)

**Comment by ztNFny** [comment]
unassigning myself then as that SAF stuff in itself and Mapsforge is too complicated for me

I'll make my latest non-SAF implementation available in https://github.com/ztNFny/cgeo/tree/issue/12210_hillshading if someone wants to pick up on that.

**Comment by eddiemuc** [comment]
I am currently bound to the variable/coordinate calculation stuff. Once that's done I can support here. But this might be a little while.

**Comment by ztNFny** [comment]
![Screenshot_1639684216](https://user-images.githubusercontent.com/1258173/146440313-34e7ee73-e056-4a15-8fc1-3c87c39ba2d6.png)
![Screenshot_1639684476](https://user-images.githubusercontent.com/1258173/146440316-232530ef-c236-4c30-ab88-d4d2ea66d0d3.png)

Implementation of Hillshading itself is pretty complete. Settings for selecting the folder (currently disabled) and adjusting the level of shading, helper methods for drawing it, strings.
Screenshots represent the latest state of https://github.com/ztNFny/cgeo/tree/issue/12210_hillshading

I'll take a look at possible downloader sources next - checking with service operators whether we could integrate them.

**Comment by devemux86** [comment]
> Mapsforge got SAF support for hillshading with https://github.com/mapsforge/mapsforge/pull/1338 (in trunk)
I'll work on integrating that in the c:geo codebase as soon as a mapsforge release becomes available

There are no immediate plans for another release (we just had one).
You can use and test the snapshot version (see the [integration guide](https://github.com/mapsforge/mapsforge/blob/master/docs/Integration.md#snapshots)).

**Comment by moving-bits** [comment]
Hi @ztNFny,

I've merged @bekuno's PR for updating the Mapsforge library to the commit including the SAF-work for hillshading, so you may continue your work on top of `master` branch.

BTW: Will hillshading also work with VTM (`UnifiedMap` OSM). If so, please extend your PR accordingly, so that we stay on level with new features for all map implementations. Thanks.

**Comment by devemux86** [comment]
> Will hillshading also work with VTM

There is no such hillshading in VTM.

VTM is an OpenGL map library, rendering (vector) data via OpenGL in real time.
Compared to Mapsforge that converts (vector) data to bitmaps before rendering.

You can use a hillshading tile source, see [MapilionMvtActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/MapilionMvtActivity.java) sample.

**Comment by usrusr** [comment]
> > If this gets implemented, can we also (optionally) show a altitude cursor like in locus?
> 
> Not with the hillshading implementation in mapsforge - is really just shading. Basically it's adding a layer with shades of grey on top of the original map.


True, the mapsforge code does not offer a convenient API for sampling the DEM at a given coordinate. But features like that locusmaps cursor altitude were definitely part of the reasoning that led to shading from a raw DEM (instead of using a pre-computed tileset that would offer acceptable perforrmance at far-out zoom leves). The place to extend would be org.mapsforge.map.layer.hills.MemoryCachingHgtReaderTileSource and its package-private HgtCache sibling, but that's all a bit convoluted because it deals in post-shading bitmaps that don't contain height data anymore and goes beyond the boundary overlap inherent in .hgt to achieve nice boundary interpolation at high zoom levels (the overlap included in hgt is enough to determine gradient or interpolate height, but insufficient to determine gradient on the out-of-tile neighbor for interpolating gradient)
